runtime/major_gc.c

```diff
@@ -267,20 +419,18 @@ Caml_inline value* mark_slice_darken(value *gray_vals_ptr,
 #endif
     if (Is_white_hd (chd)){
       ephe_list_pure = 0;
-      Hd_val (child) = Grayhd_hd (chd);
-      *gray_vals_ptr++ = child;
-      if (gray_vals_ptr >= gray_vals_end) {
-        gray_vals_cur = gray_vals_ptr;
-        realloc_gray_vals ();
-        gray_vals_ptr = gray_vals_cur;
+      Hd_val (child) = Blackhd_hd (chd);
+      if( Tag_hd(chd) < No_scan_tag ) {
+        mark_stack_push(stk, child, 0, work);
+      }
+      else {
+        *work -= 1; /* Account for header */
```

Comment:
My intention was to match the existing accounting so this is a mistake.

If my understanding is correct, changing 429 to `work -= Whsize_hd(chd);`would be sufficient in this case?

I notice we also test for No_scan_tag in `caml_darken` in this change and the original and don't add to the stack/gray_vals in that case but neither of which account that. Should we change this too?

Reply:
>If my understanding is correct, changing 429 to `work -= Whsize_hd(chd);` would be sufficient in this case?

Yes, that's what I think too.

>I notice we also test for No_scan_tag in caml_darken in this change and the original and don't add to the stack/gray_vals in that case but neither of which account that. Should we change this too?

That's a bug in the current version. It would be nice to change it, but that might get in the way of comparative benchmarking. I don't know if you intend to to much more benchmarking, so I'll let you decide.

Reply:
I'm happy to do a PR fix against trunk first, then rebase and re-benchmark after all the review changes are made on this PR against the fixed trunk.

I had originally thought there might not be much of a different for pacing if this was fixed but I could well imagine having a large array in the roots.

Reply:
Famous last words. Having done a bit more research, a fix to `caml_darken` may be complicated because of the number of different contexts it is called from e.g `caml_modify_field`.

I think the way forward is probably to have a helper function for the cases where we are called without a work balance and to credit the work done to `caml_major_work_credit`.

I'm going to avoid doing this as part of this PR though as it will change the pacing and should probably be evaluated on its own.

Reply:
That's fine with me.